dockstore-webservice/src/main/java/io/dockstore/webservice/languages/IpynbHandler.java

```diff
@@ -0,0 +1,206 @@
+    public static final Logger LOG = LoggerFactory.getLogger(IpynbHandler.class);
+
+    public static final Set<String> REES_FILES = Set.of("environment.yml", "Pipfile", "Pipfile.lock", "requirements.txt", "setup.py", "Project.toml", "REQUIRE", "install.R", "apt.txt", "DESCRIPTION", "postBuild", "start", "runtime.txt", "default.nix", "Dockerfile");
+    public static final Set<String> REES_DIRS = Set.of("/", "/binder/", "/.binder/");
+
+    @Override
+    public Version parseWorkflowContent(String filePath, String content, Set<SourceFile> sourceFiles, Version version) {
+
+        // Parse the notebook JSON.
+        JSONObject notebook;
+        try {
+            notebook = new JSONObject(content);
+        } catch (JSONException ex) {
+            LOG.error("Could not parse notebook", ex);
+            return version;
+        }
+
+        // Extract the authors from the metadata.
+        try {
+            // This code will intentionally throw a JSONException if the "authors" field is not present.
+            JSONArray jsonAuthors = notebook.getJSONObject("metadata").getJSONArray("authors");
+            Set<Author> versionAuthors = new LinkedHashSet<>();
+            for (int i = 0; i < jsonAuthors.length(); i++) {
+                JSONObject jsonAuthor = jsonAuthors.getJSONObject(i);
+                String name = jsonAuthor.optString("name", null);
+                if (name != null && name.length() > 0) {
+                    Author versionAuthor = new Author();
+                    versionAuthor.setName(name);
+                    versionAuthors.add(versionAuthor);
+                }
+                LOG.info("Notebook file contains author {}", name);
+            }
+            version.setAuthors(versionAuthors);
+        } catch (JSONException ex) {
+            LOG.warn("Could not extract notebook author information", ex);
+        }
+
+        // Extract the notebook format number from the metadata.
+        try {
+            int formatMajor = notebook.getInt("nbformat");
+            int formatMinor = notebook.getInt("nbformat_minor");
+            String format = formatMajor + "." + formatMinor;
+            sourceFiles.stream()
+                .filter(file -> file.getAbsolutePath().equals(filePath))
+                .forEach(file -> file.setTypeVersion(format));
+            version.setDescriptorTypeVersionsFromSourceFiles(sourceFiles);
+        } catch (JSONException ex) {
+            LOG.warn("Could not extract version information", ex);
+        }
+
+        return version;
+    }
+
+    @Override
+    public Map<String, SourceFile> processImports(String repositoryId, String content, Version version,
+        SourceCodeRepoInterface sourceCodeRepoInterface, String workingDirectoryForFile) {
+        Map<String, SourceFile> pathsToFiles = new HashMap<>();
+        // For each possible REES directory, list the directory contents, and if we find configuration files, read them.
```

Comment:
Try listing the folders first and then going into them if they exist (like how you do on line 107). 
This prevents an un-cachable 404